ls1intum/Artemis
# `Programming exercises`: Add manual feedback requests

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


Refer to issue #5645

### Description

Add a feature for programming exercises that are manually assessed for students to request manual feedback before the deadline.

This action is only possible, if the user already passed 100% in the automatic tests.

The feedback requests are then available for tutors/instructors in the chronological order, in which they were submitted.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with 100% passed automatic tests.

Test settings for instructors
1. Create a programming exercise
2. Set assessment to manual, and a due date. There should be a toggle "allow feedback requests"
3. Enable feedback requests and click save
4. To verify you can look into your database at exercises and see that the flag `allow_manual_feedback_requests` is set.

Test requesting the feedback
1. Get 100% at a programming exercise, where the option is enabled.
2. The `request feedback` button should be enabled then, which you can click.
3. Login as instructor and see the submission in the exercise dashboard. 

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
|programming-exercise-lifecycle.component.ts                           |       88 |   96.15 |
|exercise-details-student-actions.component.ts                         |    69.15 |    80.8 |
|assessment-warning.component.ts                                       |      100 |     100 |
|exercise-assessment-dashboard.component.ts                            |     73.6 |   93.66 |
| de.tum.in.www1.artemis.domain | 75 | 92.61 |
| de.tum.in.www1.artemis.service.SubmissionService | 77 | 94.4 |
| de.tum.in.www1.artemis.service.programming | 80 | 77.45 |
| de.tum.in.www1.artemis.web.rest.ParticipationResource | 66 | 92.19 |
| de.tum.in.www1.artemis.web.rest.ProgrammingAssessmentResource | 70 | 91.36 |



### Screenshots

![image](https://user-images.githubusercontent.com/25489463/190922077-66f7fbf8-90d1-496c-9840-3352ed1b52d2.png)

![image](https://user-images.githubusercontent.com/25489463/190922042-996e1bf6-c26e-4bb7-84bb-5ed0f7291641.png)

![Screenshot_20220929_165717](https://user-images.githubusercontent.com/25489463/193070730-b1b6ba95-323f-4bd4-9bcb-e49d0caa66ca.png)
